projects/concurrency/cache.go

```diff
@@ -0,0 +1,191 @@
+package concurrency
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+)
+
+type Node[K comparable] struct {
+	key  K
+	prev *Node[K]
+	next *Node[K]
+}
+
+type Statistics struct {
+	hitRate             float64
+	unReadEntries       int
+	averageReads        float64
+	totalReadsAndWrites int
+}
+
+type cacheEntry[K comparable, V any] struct {
+	value             V
+	timeStamp         time.Time
+	numberOfTimesRead int
+	node              *Node[K]
+}
+
+type Cache[K comparable, V any] struct {
+	entryLimit int
+
+	mu                   sync.Mutex
+	cache                map[K]cacheEntry[K, V]
+	unReadEvictedEntries int
+	totalReads           int
+	totalWrites          int
+	successfulReads      int
+	list                 *LinkedList[K]
+}
+
+type LinkedList[K comparable] struct {
+	head *Node[K]
+	tail *Node[K]
+}
+
+func NewCache[K comparable, V any](entryLimit int) Cache[K, V] {
+	return Cache[K, V]{entryLimit: entryLimit, cache: make(map[K]cacheEntry[K, V]), list: &LinkedList[K]{}}
+}
+
+func (c *Cache[K, V]) Put(key K, value V) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.totalWrites += 1
+
+	entry := cacheEntry[K, V]{value: value, timeStamp: time.Now()}
+
+	if value, ok := c.cache[key]; ok {
+		// means that it is already in the list, move to head
+		entry.numberOfTimesRead = value.numberOfTimesRead
+		c.cache[key] = entry
+		return ok
+	}
+
+	//inserting new item
+	if len(c.cache) == c.entryLimit {
+		// get key at the end of the list, delete from map and linked list
+		lastUsed := getLastRecentlyUsedEntry(c.cache)
+		if c.cache[lastUsed].numberOfTimesRead == 0 {
+			c.unReadEvictedEntries += 1
+		}
+		delete(c.cache, lastUsed)
+	}
+
+	// insert new item at the beginning of the list
+	c.cache[key] = entry
```

Comment:
I see the comment saying insert at the beginning of the list, but I only see the implementation adding to a map?